Samples/VirtualCamera/VirtualCameraMediaSource/AugmentedMediaSource.cpp

```diff
@@ -0,0 +1,1012 @@
+            }
+            else
+            {
+                return E_NOTIMPL;
+            }
+        }
+        
+
+        return hr;
+    }
+
+    IFACEMETHODIMP AugmentedMediaSource::KsMethod(
+        _In_reads_bytes_(ulMethodLength) PKSMETHOD pMethod,
+        _In_ ULONG ulMethodLength,
+        _Inout_updates_to_(ulDataLength, *pBytesReturned) LPVOID pMethodData,
+        _In_ ULONG ulDataLength,
+        _Out_ ULONG* pBytesReturned
+        )
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        // defer to source device
+        wil::com_ptr_nothrow<IKsControl> spKsControl;
+        if (SUCCEEDED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spKsControl))))
+        {
+            RETURN_IF_FAILED(spKsControl->KsMethod(pMethod, ulMethodLength, pMethodData, ulDataLength, pBytesReturned));
+        }
+        else
+        {
+            return E_NOTIMPL;
+        }
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP AugmentedMediaSource::KsEvent(
+        _In_reads_bytes_opt_(ulEventLength) PKSEVENT pEvent,
+        _In_ ULONG ulEventLength,
+        _Inout_updates_to_(ulDataLength, *pBytesReturned) LPVOID pEventData,
+        _In_ ULONG ulDataLength,
+        _Out_opt_ ULONG* pBytesReturned
+        )
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        // defer to source device
+        wil::com_ptr_nothrow<IKsControl> spKsControl;
+        if (m_spDevSource->QueryInterface(IID_PPV_ARGS(&spKsControl)))
+        {
+            RETURN_IF_FAILED(spKsControl->KsEvent(pEvent, ulEventLength, pEventData, ulDataLength, pBytesReturned));
+        }
+        else
+        {
+            return E_NOTIMPL;
+        }
+        return S_OK;
+    }
+
+    IFACEMETHODIMP AugmentedMediaSource::SetDefaultAllocator(
+        _In_  DWORD dwOutputStreamID,
+        _In_  IUnknown* pAllocator)
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        wil::com_ptr_nothrow<IMFSampleAllocatorControl> spAllocatorControl;
+
+        RETURN_IF_FAILED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spAllocatorControl)));
+        RETURN_IF_FAILED(spAllocatorControl->SetDefaultAllocator(m_streamList[0]->m_dwDevSourceStreamIdentifier, pAllocator));
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP AugmentedMediaSource::GetAllocatorUsage(
+        _In_  DWORD dwOutputStreamID,
+        _Out_  DWORD* pdwInputStreamID,
+        _Out_  MFSampleAllocatorUsage* peUsage)
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        RETURN_HR_IF_NULL(E_POINTER, pdwInputStreamID);
+        RETURN_HR_IF_NULL(E_POINTER, peUsage);
+
+        wil::com_ptr_nothrow<IMFSampleAllocatorControl> spAllocatorControl;
+        if (SUCCEEDED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spAllocatorControl))))
```

Comment:
if the processSample is going to do inplace transform then it should report MFSampleAllocatorUsage_DoesNotAllocate.  However if it will process and place into a new sample it should report MFSampleAllocatorUsage_UsesProvidedAllocator